File: converter.py
import os


def convertPlaylist(
    inputPath,
    outputPath,
    basePath=None,
    locationMode="keep"  # keep | custom
):

    if not inputPath.lower().endswith((".m3u", ".m3u8")):
        raise ValueError("Unsupported file type")

    inputExt = os.path.splitext(inputPath)[1].lower()

    if inputExt == ".m3u":
        if not basePath:
            raise ValueError("Base path required for M3U → M3U8 conversion")
        _convertM3uToM3u8(inputPath, outputPath, basePath)

    elif inputExt == ".m3u8":
        _convertM3u8ToM3u(inputPath, outputPath, locationMode, basePath)


# ================= M3U to M3U8 =================

def _convertM3uToM3u8(inputPath, outputPath, basePath):

    basePath = basePath.rstrip("/\\").replace("\\", "/")

    outputName = os.path.splitext(os.path.basename(outputPath))[0]

    with open(inputPath, "r", encoding="utf-8", errors="ignore") as infile, \
         open(outputPath, "w", encoding="utf-8") as outfile:

        outfile.write("#EXTM3U\n")
        outfile.write(f"#{outputName}.m3u8\n")

        for line in infile:
            line = line.strip()

            if not line or line.startswith("#"):
                continue

            if line.startswith("Music/"):
                line = line[len("Music/"):]

            line = line.replace("\\", "/")

            fullPath = f"{basePath}/{line}"
            outfile.write(fullPath + "\n")

# ================= M3U8 to M3U =================

def _convertM3u8ToM3u(inputPath, outputPath, locationMode, customBase):

    customBaseNormalized = None

    if locationMode == "custom":
        if not customBase:
            raise ValueError("Custom base path required")
        customBaseNormalized = customBase.rstrip("/\\").replace("\\", "/")

    with open(inputPath, "r", encoding="utf-8", errors="ignore") as infile, \
         open(outputPath, "w", encoding="utf-8") as outfile:

        for line in infile:
            stripped = line.strip()

            if not stripped or stripped.startswith("#"):
                continue

            originalPath = stripped.replace("\\", "/")
            filename = os.path.basename(originalPath)

            if locationMode == "keep":
                outfile.write(originalPath + "\n")

            elif locationMode == "custom":
                newPath = f"{customBaseNormalized}/{filename}"
                outfile.write(newPath + "\n")

File: src/Converter.cpp
#include "Converter.h"
#include "Logger.h"
#include <QFile>
#include <QTextStream>
#include <QFileInfo>

Q_LOGGING_CATEGORY(lcConverter, "le.converter")

namespace LE {

void Converter::convert(const ConversionParams& params)
{
    qCInfo(lcConverter) << "Conversion start:" << params.inputPath << "->" << params.outputPath;

    if (!params.inputPath.endsWith(".m3u", Qt::CaseInsensitive) &&
        !params.inputPath.endsWith(".m3u8", Qt::CaseInsensitive))
    {
        throw std::runtime_error("Unsupported file type. Expected .m3u or .m3u8.");
    }

    if (params.inputPath.endsWith(".m3u", Qt::CaseInsensitive)) {
        convertM3uToM3u8(params);
    } else {
        convertM3u8ToM3u(params);
    }

    qCInfo(lcConverter) << "Conversion complete:" << params.outputPath;
}

// ─── M3U → M3U8 ────────────────────────────────────────────────────────────

void Converter::convertM3uToM3u8(const ConversionParams& params)
{
    if (params.basePath.isEmpty()) {
        throw std::runtime_error("Base path is required for M3U → M3U8 conversion.");
    }

    const QString base = normalizePath(params.basePath);
    const QString outputName = QFileInfo(params.outputPath).completeBaseName();

    QFile inFile(params.inputPath);
    if (!inFile.open(QIODevice::ReadOnly | QIODevice::Text)) {
        qCCritical(lcConverter) << "Failed to open input file:" << params.inputPath;
        throw std::runtime_error("Cannot open input file: " + params.inputPath.toStdString());
    }

    QFile outFile(params.outputPath);
    if (!outFile.open(QIODevice::WriteOnly | QIODevice::Text | QIODevice::Truncate)) {
        qCCritical(lcConverter) << "Failed to open output file:" << params.outputPath;
        throw std::runtime_error("Cannot open output file: " + params.outputPath.toStdString());
    }

    QTextStream in(&inFile);
    QTextStream out(&outFile);
    in.setEncoding(QStringConverter::Utf8);
    out.setEncoding(QStringConverter::Utf8);

    out << "#EXTM3U\n";
    out << "#" << outputName << ".m3u8\n";

    while (!in.atEnd()) {
        QString line = in.readLine().trimmed();

        if (line.isEmpty() || line.startsWith('#')) {
            continue;
        }

        line = stripLeadingMusicPrefix(line);
        line = normalizePath(line);

        const QString fullPath = base + '/' + line;
        out << fullPath << '\n';
    }

    qCDebug(lcConverter) << "M3U→M3U8 written to:" << params.outputPath;
}

// ─── M3U8 → M3U ────────────────────────────────────────────────────────────

void Converter::convertM3u8ToM3u(const ConversionParams& params)
{
    QString customBase;

    if (params.locationMode == LocationMode::Custom) {
        if (params.basePath.isEmpty()) {
            throw std::runtime_error("Custom base path is required for custom location mode.");
        }
        customBase = normalizePath(params.basePath);
    }

    QFile inFile(params.inputPath);
    if (!inFile.open(QIODevice::ReadOnly | QIODevice::Text)) {
        qCCritical(lcConverter) << "Failed to open input file:" << params.inputPath;
        throw std::runtime_error("Cannot open input file: " + params.inputPath.toStdString());
    }

    QFile outFile(params.outputPath);
    if (!outFile.open(QIODevice::WriteOnly | QIODevice::Text | QIODevice::Truncate)) {
        qCCritical(lcConverter) << "Failed to open output file:" << params.outputPath;
        throw std::runtime_error("Cannot open output file: " + params.outputPath.toStdString());
    }

    QTextStream in(&inFile);
    QTextStream out(&outFile);
    in.setEncoding(QStringConverter::Utf8);
    out.setEncoding(QStringConverter::Utf8);

    while (!in.atEnd()) {
        QString line = in.readLine().trimmed();

        if (line.isEmpty() || line.startsWith('#')) {
            continue;
        }

        const QString originalPath = normalizePath(line);

        if (params.locationMode == LocationMode::Keep) {
            out << originalPath << '\n';
        } else {
            const QString filename = QFileInfo(originalPath).fileName();
            out << customBase << '/' << filename << '\n';
        }
    }

    qCDebug(lcConverter) << "M3U8→M3U written to:" << params.outputPath;
}

// ─── Helpers ────────────────────────────────────────────────────────────────

QString Converter::normalizePath(const QString& path)
{
    QString normalized = path;
    normalized.replace('\\', '/');

    while (normalized.endsWith('/')) {
        normalized.chop(1);
    }

    return normalized;
}

QString Converter::stripLeadingMusicPrefix(const QString& line)
{
    constexpr QLatin1StringView prefix{"Music/"};
    if (line.startsWith(prefix)) {
        return line.sliced(prefix.size());
    }
    return line;
}

} // namespace LE